src/util/FormUtils.js

```diff
@@ -478,4 +478,103 @@ export default class FormUtils {
       recursiveTraverse(q);
     });
   }
+
+  /**
+   * Performs a dfs on a set of questions and executes callback functions
+   * when entering and leaving each question.
+   * @param {Array<Object>} questions - The array of questions to traverse.
+   * @param {function} onEnterQuestion - The callback function to execute when entering a question.
+   * @param {function} onLeaveQuestion - The callback function to execute when leaving a question.
+   */
+  static dfsTraverseQuestionTreeExtended(
+    questions,
+    onEnterQuestion,
+    onLeaveQuestion
+  ) {
+    /**
+     * Recursively traverses the questions and executes the appropriate callback functions.
+     * @param {Object} question - The question to traverse.
+     */
+    function recursiveTraverse(question) {
+      onEnterQuestion(question);
+      Utils.asArray(question[Constants.HAS_SUBQUESTION]).forEach((q) => {
+        recursiveTraverse(q);
+      });
+      onLeaveQuestion(question);
+    }
+
+    questions.forEach((q) => {
+      recursiveTraverse(q);
+    });
+  }
+
+  static getFormSpecification(questions) {
+    let level = -1;
+    const indentation = "    ";
+    const propertyIndentation = "..";
+    let output = "";
+
+    function onEnterQuestion(q) {
+      level += 1;
+      const ind = indentation.repeat(level);
+      let requiredValue = "";
+      if (
+        q[Constants.REQUIRES_ANSWER] &&
+        !q[Constants.USED_ONLY_FOR_COMPLETENESS]
+      ) {
+        requiredValue =
+          ind +
+          propertyIndentation +
+          "required: " +
```

Comment:
I prefer to have 3 values of one property as suggested, e.g.:
`requirement: NONE`
`requirement: REQUIRED_TO_SAVE_FORM
`requirement: REQUIRED_TO_COMPLETE_FORM`



My motivation here is that it is easier to read/search and understand for the enduser.

Reply:
@LaChope